File: CountryProject/main/Python/RSAFull ARIMA Modeling with Saving Summaries.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from datetime import datetime
from statsmodels.tsa.seasonal import seasonal_decompose
from statsmodels.tsa.arima.model import ARIMA

# 1. Data Loading and Preparation
class EurostatData:
    def __init__(self, filepath):
        self.df = pd.read_excel(filepath)
        self.df['time'] = pd.to_datetime(self.df['time'])
        self.df = self.df[self.df['time'] >= '2010-01-01']

# 2. Decomposition Analysis
class DecompositionAnalysis:
    def __init__(self, data):
        self.data = data
        self.countries = data['geo'].unique()

    def perform_decomposition(self, model='additive'):
        decomposition_results = {}
        for country in self.countries:
            country_data = self.data[self.data['geo'] == country].set_index('time')['values'].dropna()
            if len(country_data) < 24:
                continue
            decomposition = seasonal_decompose(country_data, model=model, period=12)
            decomposition_results[country] = decomposition
            # Save decomposition components for each country
            result_df = pd.DataFrame({
                "trend": decomposition.trend,
                "seasonal": decomposition.seasonal,
                "residual": decomposition.resid
            })
            result_df.to_csv(f"{country}_decomposition_{model}.csv", index_label="date")
        return decomposition_results

# 3. Outlier Removal
class OutlierRemoval:
    def __init__(self, decomposition_results, start_date="2020-01-01", end_date="2021-12-31"):
        self.decomposition_results = decomposition_results
        self.start_date = datetime.strptime(start_date, "%Y-%m-%d")
        self.end_date = datetime.strptime(end_date, "%Y-%m-%d")
        self.cleaned_results = {}

    def remove_outliers(self):
        for country, decomposition in self.decomposition_results.items():
            residuals = decomposition.resid.dropna()
            residuals_df = residuals.to_frame(name='residuals')
            residuals_df['date'] = residuals_df.index

            covid_period = (residuals_df['date'] >= self.start_date) & (residuals_df['date'] <= self.end_date)
            residuals_df = residuals_df[~covid_period]
            residuals_df.to_csv(f"{country}_cleaned_residuals.csv", index=False)
            self.cleaned_results[country] = residuals_df.set_index('date')['residuals']
        return self.cleaned_results

# 4. RSAFull ARIMA Modeling with Saving Summaries
class RSAFullModel:
    def __init__(self, cleaned_residuals):
        self.cleaned_residuals = cleaned_residuals
        self.models = {}

    def fit_arima_with_regressors(self, order=(1, 1, 1), seasonal_order=(0, 1, 1, 12)):
        for country, residuals in self.cleaned_residuals.items():
            residuals_df = residuals.to_frame(name='residuals')
            residuals_df['working_day'] = np.where(residuals_df.index.dayofweek < 5, 1, 0)
            try:
                model = ARIMA(residuals_df['residuals'], order=order, seasonal_order=seasonal_order, exog=residuals_df[['working_day']])
                fitted_model = model.fit()
                self.models[country] = fitted_model
                
                # Save model summary to a text file
                with open(f"{country}_rsafull_model_summary.txt", "w") as f:
                    f.write(f"RSAFull ARIMA Model Summary for {country}\n")
                    f.write(fitted_model.summary().as_text())
            except Exception as e:
                print(f"Could not fit RSAFull ARIMA model for {country}: {e}")

# Final Script Usage
data = EurostatData("./dane_eurostat.xlsx")
analysis = DecompositionAnalysis(data.df)
decomposition_results = analysis.perform_decomposition(model='additive')

outlier_removal = OutlierRemoval(decomposition_results, start_date="2020-01-01", end_date="2021-12-31")
cleaned_residuals = outlier_removal.remove_outliers()

rsafull_model = RSAFullModel(cleaned_residuals)
rsafull_model.fit_arima_with_regressors(order=(1, 1, 1), seasonal_order=(0, 1, 1, 12))

File: CountryProject/main/R/RSAFull ARIMA Modeling with Saving Summaries.R
library(tidyverse)
library(forecast)
library(RJDemetra)
library(lubridate)
library(zoo)

# 1. Data Loading and Preparation
load_data <- function(filepath) {
  data <- readxl::read_excel(filepath) %>%
    mutate(time = as.Date(time)) %>%
    filter(time >= as.Date("2010-01-01"))
  return(data)
}

# 2. Decomposition Analysis with Saving
perform_decomposition <- function(data, model = "additive") {
  decomposition_results <- list()
  countries <- unique(data$geo)
  
  for (country in countries) {
    country_data <- data %>% filter(geo == country) %>% select(time, values) %>%
      arrange(time)
    ts_data <- ts(country_data$values, frequency = 12, start = c(year(min(country_data$time)), month(min(country_data$time))))
    
    decomposition <- decompose(ts_data, type = model)
    decomposition_results[[country]] <- decomposition

    # Save decomposition components
    result_df <- data.frame(
      date = as.Date(time(ts_data)),
      trend = decomposition$trend,
      seasonal = decomposition$seasonal,
      residual = decomposition$random
    )
    write.csv(result_df, paste0(country, "_decomposition_", model, ".csv"), row.names = FALSE)
  }
  return(decomposition_results)
}

# 3. Outlier Removal with Saving
remove_outliers <- function(decomposition_results, start_date = "2020-01-01", end_date = "2021-12-31") {
  cleaned_results <- list()
  covid_period <- seq(as.Date(start_date), as.Date(end_date), by = "days")
  
  for (country in names(decomposition_results)) {
    residuals <- decomposition_results[[country]]$random
    residuals_time <- time(residuals)
    
    non_covid_residuals <- residuals[!residuals_time %in% covid_period]
    result_df <- data.frame(date = as.Date(time(non_covid_residuals)), residuals = non_covid_residuals)
    write.csv(result_df, paste0(country, "_cleaned_residuals.csv"), row.names = FALSE)
    cleaned_results[[country]] <- non_covid_residuals
  }
  return(cleaned_results)
}

# 4. RSAFull ARIMA Modeling with Saving Summaries
fit_rsafull_arima <- function(cleaned_residuals, order = c(1, 1, 1), seasonal_order = c(0, 1, 1, 12)) {
  models <- list()
  
  for (country in names(cleaned_residuals)) {
    residuals <- cleaned_residuals[[country]]
    ts_data <- ts(residuals, frequency = 12)
    working_days <- ifelse(weekdays(as.Date(time(ts_data))) %in% c("Saturday", "Sunday"), 0, 1)
    
    tryCatch({
      model <- auto.arima(ts_data, xreg = working_days, seasonal = TRUE, approximation = FALSE, stepwise = FALSE)
      models[[country]] <- model
      
      # Save model summary to text file
      sink(paste0(country, "_rsafull_model_summary.txt"))
      print(summary(model))
      sink()
    }, error = function(e) {
      message(paste("Failed to fit RSAFull ARIMA model for", country, ":", e$message))
    })
  }
  return(models)
}

# Final Script Usage
data <- load_data("dane_eurostat.xlsx")
decomposition_results <- perform_decomposition(data, model = "additive")
cleaned_residuals <- remove_outliers(decomposition_results, start_date = "2020-01-01", end_date = "2021-12-31")
rsafull_models <- fit_rsafull_arima(cleaned_residuals, order = c(1, 1, 1), seasonal



